### DISCUSSION TITLE
Is there some way to preserve the state of individual nested machines when superstate changes?

### DISCUSSION THREAD

**Discussion by denisstasyev** [statelyai/xstate#2863] [Q&A] [answered]
Hi, is there some way not to re-create nested machines when transitioning between superstates (in other words, keep the state of nested machines when changing superstates)?

For example, let's say I have a state machine (let's call it a grocery store) with two superstates:
- Sale of vegetables :tomato: and bread :bread:
- Sale of fruits :tangerine: and bread :bread:

Each superstate contains two nested machines (respectively):
- Shelf with vegetables :tomato: | shelf with bread :bread:
- Fruit shelf :tangerine: | bread shelf :bread:

As I understand it, when using the concept of nested machines, I cannot switch between the described superstates without recreating the nested machine for the bread shelf.

But what if I need to maintain the internal state of the bread shelf when the top-level superstate changes? If there are any solutions that are not based on the use of the Actor model?

I suspect that a similar problem can be solved by writing some additional parameters into the context, but I have not come up with an elegant solution. Therefore, I am asking the community for help :)

P.S. I will be glad to hear fresh ideas so that fruits and vegetables do not go bad!

**Comment by davidkpiano** [comment]
With XState, there is a somewhat-implicit **root state**, and this state is active for the duration of the machine. This is a state where you can invoke other machines that need to also be active for the duration of the machine:

```js
const machine = createMachine({
  // this is the root-level!
  invoke: [
    { src: veggieMachine, id: 'veggies' },
    { src: fruitMachine, id: 'fruits' }
  ],
  initial: 'sellingVeggies',
  states: {
    sellingVeggies: {/* talk to the 'veggies' actor */},
    sellingFruits: {/* talk to the 'fruits' actor */}
  }
});
```

**Reply by denisstasyev** [reply]
Thanks for the idea, but as I understand it, it will turn out all the same as actors: that is, these involved machines will live for all superstates.

Is it possible to create nested machines that will exist only for some superstates (and can only survive in a limited set of transitions between superstates)? Or is it just "normal" actors ([XState Docs of Actors](https://xstate.js.org/docs/guides/actors.html)) with dynamic spawning?

**Comment by denisstasyev** [comment, marked as answer]
Thanks very much @mogsie for a fresh look that the described problem can be solved with a slightly different look at the design of nested machines. The solution is described in [this comment].(https://github.com/statecharts/statecharts/discussions/30#discussioncomment-1773453)
